commandbased/commandbasedrobot.py

```diff
@@ -0,0 +1,116 @@
+import hal
+
+from wpilib.iterativerobot import IterativeRobot
+from wpilib.command.scheduler import Scheduler
+from wpilib.driverstation import DriverStation
+from wpilib.livewindow import LiveWindow
+
+
+class CommandBasedRobot(IterativeRobot):
+    '''
+    The base class for a Command-Based Robot. To use, instantiate commands and
+    trigger them.
+    '''
+
+
+    def startCompetition(self):
+        """
+        Provide an alternate "main loop" via startCompetition(). Rewritten
+        from IterativeRobot for readability and to initialize scheduler.
+        """
+
+        hal.HALReport(hal.HALUsageReporting.kResourceType_Framework,
+                      hal.HALUsageReporting.kFramework_Iterative)
+
+        self.scheduler = Scheduler.getInstance()
+        self.robotInit()
+
+        # Tell the DS that the robot is ready to be enabled
+        hal.HALNetworkCommunicationObserveUserProgramStarting()
+
+        # loop forever, calling the appropriate mode-dependent function
+        while True:
+            if self.ds.isDisabled():
+                self.disabledInit()
+                while self.ds.isDisabled():
+                    hal.HALNetworkCommunicationObserveUserProgramDisabled()
+                    self.__runPeriodic(self.disabledPeriodic)
+
+            elif self.ds.isAutonomous():
+                self.autonomousInit()
+                while self.ds.isAutonomous():
+                    hal.HALNetworkCommunicationObserveUserProgramAutonomous()
+                    self.__runPeriodic(self.autonomousPeriodic)
+
+            elif self.ds.isTest():
+                LiveWindow.setEnabled(True)
+
+                self.testInit()
+                while self.ds.isTest():
+                    hal.HALNetworkCommunicationObserveUserProgramTest()
+                    self.__runPeriodic(self.testPeriodic)
+
+                LiveWindow.setEnabled(False)
+
+            else:
+                self.teleopInit()
+                # isOperatorControl checks "not autonomous or test", so we need
+                # to check isEnabled as well, since otherwise it will continue
+                # looping while disabled.
+                while self.ds.isEnabled() and self.ds.isOperatorControl():
+                    hal.HALNetworkCommunicationObserveUserProgramTeleop()
+                    self.__runPeriodic(self.teleopPeriodic)
+
+
+    def __runPeriodic(self, periodicFunction):
+        '''
+        Run the correct "periodic" function and then wait for new data from the
+        driver station.
+        '''
+
+        periodicFunction()
+
+        while not self.ds.isNewControlData():
```

Comment:
It wasn't obvious from the code that the two are equivalent. Here are their definitions.

```
def waitForData(self, timeout = None):
        with self.dataSem:
            self.dataSem.wait(timeout)
```

```
def isNewControlData(self):
        with self.mutex:
            result = self.newControlData
            self.newControlData = False
            return result
```

They seem to be doing very similar things, but they don't reference any of the same variables (although `dataSem` is created from `mutex`). When `waitForData()` returns is `isNewControlData()` guaranteed to be `True`. If so, we can remove the loop.